docker/tarfile/dest.go

```diff
@@ -122,18 +130,49 @@ func (d *Destination) PutBlob(stream io.Reader, inputInfo types.BlobInfo) (types
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
+		if inputInfo.Digest == "" {
+			digester := digest.Canonical.Digester()
+			_, err = io.Copy(digester.Hash(), streamCopy)
+			if err != nil {
+				return types.BlobInfo{}, err
+			}
+			_, err = streamCopy.Seek(0, os.SEEK_SET)
+			if err != nil {
+				return types.BlobInfo{}, err
+			}
+			inputInfo.Digest = digester.Digest()
```

Comment:
It would probably be cheaper to either compute the digest always, or use `io.TeeReader`/`io.MultiWriter`, so that we make as few passes over the data as possible.